### PR TITLE
Add unit tests

### LINKED ISSUE #1: Add unit tests
- [x] DsLineSocket
   - [x] is Connected/Disconnected
   - [x] stream
   - [x] requestAll
      - [x] when connected
      - [x] when not connected
   - [x] send
   - [x] close
      - [x] isConnected = false, 
      - [x] exit from _listenSocket()
- [x] JdsLine
   - [x] stream
   - [x] requestAll
      - [x] when connected
      - [x] when not connected
   - [x] send
   - [x] close
- [x] DsClientReal
   - [x] stream (30 subscriptions)
      - подписаться
      - отписаться
      - повторять 10 раз
      - проверить что подписки освобождены

### PR DESCRIPTION
**Pull request by Minyewoo**

### CODE REVIEW

File: test/unit/helpers.dart [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      (dataPoint) => utf8.encode(dataPoint.toJson())
+        .toList()
+        ..add(Jds.endOfTransmission),
+    )
+    .expand((element) => element)
+    .toList(),
+  );
+}
+
+bool compareDataPoints(DsDataPoint first, DsDataPoint second) {
```
Comment (a-givertzman) [inline test/unit/helpers.dart]: let's implement it inside DsDataPoint

issue added: [DsDataPoint | operator=](https://github.com/a-givertzman/hmi_core/issues/6)

Comment (a-givertzman) [reply test/unit/helpers.dart:38]: issue added: #6

File: test/unit/core/ds_line_socket_test.dart [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+        await socketServer.close();
+        await clientSocket.close();
+
+        await Future.delayed(const Duration(milliseconds: 100));
+
+        expect(lineSocket.isConnected, false);
+      });
+
+      test('requestAll', () async {
```
Comment (a-givertzman) [inline test/unit/core/ds_line_socket_test.dart]: - [x] rename into requestAll when isConnected = true
- [x] add requestAll when isConnected = false

**Comment by a-givertzman** [comment]
- [x] divide each test into separate file

File: test/unit/core/ds_line_socket_test.dart [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        
+        const sendRepetitions = 3;
+        final testMessage = 'test';
+
+        for(var i = 0; i < sendRepetitions; i++) {
+          clientSocket.write(testMessage);
+          await Future.delayed(const Duration(milliseconds: 100));
+        }
+
+        final targetData = List.generate(sendRepetitions, (index) => testMessage);
```
Comment (a-givertzman) [inline test/unit/core/ds_line_socket_test.dart]: remove it, not reliable

File: test/unit/core/ds_line_socket_test.dart [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        lineSocketSubscription = lineSocket.stream.listen((event) { receivedData.add(utf8.decode(event)); });
+        final clientSocket = await socketServer.first;
+        
+        await Future.delayed(const Duration(milliseconds: 100));
+        
+        const sendRepetitions = 3;
+        final testMessage = 'test';
+
+        for(var i = 0; i < sendRepetitions; i++) {
+          clientSocket.write(testMessage);
```
Comment (a-givertzman) [inline test/unit/core/ds_line_socket_test.dart]: add here targetData.add(testMessage)

File: test/unit/protocols/jds_line_test.dart [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+        final receivedEvents = <DsDataPoint>[];
+        line.stream.listen((event) { receivedEvents.add(event); });
+
+        await Future.delayed(const Duration(milliseconds: 100));
+        
+        expect(compareDataPointCollections(receivedEvents, testDataPoints), true);
+      });
+
+      test('requestAll', () async {
```
Comment (a-givertzman) [inline test/unit/protocols/jds_line_test.dart]: was sent to the server the specific requestAll command  must be verified first of all:
DsCommand(
      dsClass: DsDataClass.requestAll,
      type: DsDataType.bool,
      path: '',
      name: '',
      value: 1,
      status: DsStatus.ok,
      timestamp: DsTimeStamp.now(),
    )

File: test/unit/protocols/jds_line_test.dart [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+        await Future.delayed(const Duration(milliseconds: 100));
+        final result = await line.requestAll();
+
+        await Future.delayed(const Duration(milliseconds: 100));
+        expect(result.data, true);
+        expect(receivedEvents.every((event) => event.type == DsDataType.bool && event.value == 1), true);
+      });
+
+      test('send', () async {
```
Comment (a-givertzman) [inline test/unit/protocols/jds_line_test.dart]: was sent to the server the pool of various command types must be verified first of all:

**Review by a-givertzman — COMMENTED** [review]
Fix all commented points before merging

File: test/unit/protocols/jds_line_test.dart (hunk 5 above)
Comment (Minyewoo) [reply test/unit/protocols/jds_line_test.dart]: Done:
https://github.com/a-givertzman/hmi_networking/pull/3/files#diff-2f289a8438c1fbb1711f3abbd59844e18cb9702b03b49700386d3198685598ae

File: test/unit/protocols/jds_line_test.dart (hunk 6 above)
Comment (Minyewoo) [reply test/unit/protocols/jds_line_test.dart]: Done:
https://github.com/a-givertzman/hmi_networking/pull/3/files#diff-863531f550408fda10395d04f67ae07424fd77925eda1eb4f4e441ae8e18fdb2

File: test/unit/core/ds_line_socket_test.dart (hunk 4 above)
Comment (Minyewoo) [reply test/unit/core/ds_line_socket_test.dart]: Done:
https://github.com/a-givertzman/hmi_networking/pull/3/files#diff-f7ffc4c31bd14c93001712cc4b0d48993abcecf1fbf7ddfcfb61c55b2eb9ee51

File: test/unit/core/ds_line_socket_test.dart (hunk 3 above)
Comment (Minyewoo) [reply test/unit/core/ds_line_socket_test.dart]: Done:
https://github.com/a-givertzman/hmi_networking/pull/3/files#diff-f7ffc4c31bd14c93001712cc4b0d48993abcecf1fbf7ddfcfb61c55b2eb9ee51

File: test/unit/core/ds_line_socket_test.dart (hunk 2 above)
Comment (Minyewoo) [reply test/unit/core/ds_line_socket_test.dart]: Done:
https://github.com/a-givertzman/hmi_networking/pull/3/files#diff-2f289a8438c1fbb1711f3abbd59844e18cb9702b03b49700386d3198685598ae

https://github.com/a-givertzman/hmi_networking/pull/3/files#diff-e560f30f34da2a86546f4116fde3fb6bad2223764c514bc7604de88ffd2a52ab

File: test/unit/helpers.dart (hunk 1 above)
Comment (Minyewoo) [reply test/unit/helpers.dart]: Implemented in: https://github.com/a-givertzman/hmi_core/pull/11/files#diff-7a6ffd8685914ed6994bae425ab25cc79a02a60938f4c1f96861f0e9f5861de6

Old code removed: https://github.com/a-givertzman/hmi_networking/pull/3/commits/03d59003b40c1dfed3b08297a6cc116f8d33a01d#diff-7bfa8602f509100a884256af3c31a4205f042a8ea23df0900a1a06450f9c21ee

**Pull request merged**